helix-lsp/src/snippet.rs

```diff
@@ -0,0 +1,521 @@
+                ntabstops[len_1].push((o1, o2));
+            } else {
+                prev = Some(tabstop);
+                ntabstops.push(smallvec![(o1, o2)]);
+            }
+        }
+    }
+
+    if let Some(first) = ntabstops.first() {
+        let cursor_offset = insert.chars().count() as i128 - (end_offset - start_offset);
+        let mut extra_offset = start_offset;
+        transaction.with_selection(selection.clone().transform_iter(|range| {
+            let cursor = range.cursor(text);
+            let iter = first.iter().map(move |first| {
+                Range::new(
+                    (cursor as i128 + first.0 as i128 + extra_offset) as usize,
+                    (cursor as i128 + first.1 as i128 + extra_offset) as usize,
+                )
+            });
+            extra_offset += cursor_offset;
+            iter
+        }))
+    } else {
+        transaction
+    }
+}
+
+mod parser {
+    use helix_parsec::*;
+
+    use super::{CaseChange, FormatItem, Regex, Snippet, SnippetElement};
+
+    /*
+    https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#snippet_syntax
+
+        any         ::= tabstop | placeholder | choice | variable | text
+        tabstop     ::= '$' int | '${' int '}'
+        placeholder ::= '${' int ':' any '}'
+        choice      ::= '${' int '|' text (',' text)* '|}'
+        variable    ::= '$' var | '${' var }'
+                        | '${' var ':' any '}'
+                        | '${' var '/' regex '/' (format | text)+ '/' options '}'
+        format      ::= '$' int | '${' int '}'
+                        | '${' int ':' '/upcase' | '/downcase' | '/capitalize' '}'
+                        | '${' int ':+' if '}'
+                        | '${' int ':?' if ':' else '}'
+                        | '${' int ':-' else '}' | '${' int ':' else '}'
+        regex       ::= Regular Expression value (ctor-string)
+        options     ::= Regular Expression option (ctor-options)
+        var         ::= [_a-zA-Z] [_a-zA-Z0-9]*
+        int         ::= [0-9]+
+        text        ::= .*
+        if          ::= text
+        else        ::= text
+    */
+
+    fn var<'a>() -> impl Parser<'a, Output = &'a str> {
+        // var = [_a-zA-Z][_a-zA-Z0-9]*
+        move |input: &'a str| match input
+            .char_indices()
+            .take_while(|(p, c)| {
+                *c == '_'
+                    || if *p == 0 {
+                        c.is_ascii_alphabetic()
+                    } else {
+                        c.is_ascii_alphanumeric()
+                    }
+            })
+            .last()
+        {
+            Some((index, c)) if index >= 1 => {
+                let index = index + c.len_utf8();
+                Ok((&input[index..], &input[0..index]))
+            }
+            _ => Err(input),
+        }
+    }
+
+    fn txt<'a>() -> impl Parser<'a, Output = &'a str> {
+        take_while(|c| c != '$')
+    }
```

Comment:
I would prefer that this rule is named `text` so that it matches that part of the EBNF from the spec. The current `text` rule can be inlined into `anything`:

```rust
let text = map(text(), SnippetElement::Text);
choice!(tabstop(), placeholder(), choice(), variable(), text)
```